engines/dragons/dragons.cpp

```diff
@@ -0,0 +1,1452 @@
+	_actorManager = new ActorManager(actorResourceLoader);
+	_scriptOpcodes = new ScriptOpcodes(this, _dragonFLG);
+	_backgroundResourceLoader = new BackgroundResourceLoader(_bigfileArchive, _dragonRMS);
+	_scene = new Scene(this, _screen, _scriptOpcodes, _actorManager, _dragonRMS, _dragonINIResource, _backgroundResourceLoader);
+
+	_sound = new SoundManager(this, _bigfileArchive, _dragonRMS);
+
+	if (ConfMan.hasKey("save_slot")) {
+		loadGameState(ConfMan.getInt("save_slot"));
+	} else {
+		loadScene(0);
+	}
+
+	_scene->draw();
+	_screen->updateScreen();
+
+	gameLoop();
+
+	delete _scene;
+	delete _actorManager;
+	delete _backgroundResourceLoader;
+	delete _dragonFLG;
+	delete _dragonIMG;
+	delete _dragonRMS;
+	delete _dragonVAR;
+	delete _fontManager;
+	delete _bigfileArchive;
+	delete _screen;
+	delete _sound;
+
+	debug("Ok");
+	return Common::kNoError;
+}
+
+uint16 DragonsEngine::ipt_img_file_related()
+{
+	DragonINI *ini;
+    DragonINI *flicker = _dragonINIResource->getFlickerRecord();
+	assert(flicker);
+
+	int16 tileX = flicker->actor->x_pos / 32;
+	int16 tileY = flicker->actor->y_pos / 8;
+
+	for (int i=0;i < _dragonINIResource->totalRecords(); i++) {
+		ini = getINI(i);
+		if ((ini->sceneId == getCurrentSceneId()) && (ini->field_1a_flags_maybe == 0)) {
+			IMG *img = _dragonIMG->getIMG(ini->field_2);
+			if ((img->x <= tileX) && (((tileX <= img->x + img->w && (img->y <= tileY)) && (tileY <= img->y + img->h)))) {
+				return i + 1;
+			}
+		}
+	}
+	return 0;
+}
+
+void DragonsEngine::gameLoop()
+{
+	uint uVar3;
+	uint actorId;
+	uint16 uVar6;
+	uint16 uVar7;
+	uint actorId_00;
+	uint16 sequenceId;
+	DragonINI *pDVar8;
+
+	_cursor->_cursorActivationSeqOffset = 0;
+	bit_flags_8006fbd8 = 0;
+	_counter = 0;
+	setFlags(ENGINE_FLAG_8);
+	actorId = 0;
+
+	while (!shouldQuit()) {
+		_scene->draw();
+		_screen->updateScreen();
+		wait();
+		updateHandler();
+		updateEvents();
+		runSceneUpdaterFunction(); //TODO is this the right place for this logic?
+
+		if (getCurrentSceneId() != 2) {
+			_sceneId1 = getCurrentSceneId();
+		}
+
+		_counter++;
+		if (0x4af < _counter) {
+			pDVar8 = _dragonINIResource->getFlickerRecord();
+			if (pDVar8->actor->resourceID == 0xe) {
+				pDVar8->actor->_sequenceID2 = 2;
+				pDVar8->field_20_actor_field_14 = 2;
+				if (getINI(0xc2)->field_1e == 1) {
+					sequenceId = 0x30;
+				}
+				else {
+					sequenceId = 2;
+				}
+				pDVar8->actor->updateSequence(sequenceId);
+				_counter = 0;
+				setFlags(ENGINE_FLAG_80000000);
+			}
+		}
+		if (isFlagSet(ENGINE_FLAG_80000000)
+			&& _dragonINIResource->getFlickerRecord()->actor->isFlagSet(ACTOR_FLAG_4)) {
+			_counter = 0;
+			clearFlags(ENGINE_FLAG_80000000);
+		}
+		if (bit_flags_8006fbd8 == 0) {
+			setFlags(ENGINE_FLAG_8);
+		}
+		if (_dragonINIResource->getFlickerRecord()->sceneId == getCurrentSceneId()) {
+			uVar3 = ipt_img_file_related();
+			actorId_00 = uVar3 & 0xffff;
+			if (actorId_00 == 0) goto LAB_80026d34;
```

Comment:
omg.

Reply:
yeah there's a fair bit to clean up here.